src/ast/passes/semantic_analyser.cpp

```diff
@@ -1463,30 +1463,78 @@ Probe *SemanticAnalyser::get_probe_from_scope(Scope *scope,
 
 void SemanticAnalyser::visit(Map &map)
 {
-  MapKey key;
+  SizedType key = CreateNone();
+  bool is_tuple_ast = false;
+
+  if (map.key_expr) {
+    Visit(map.key_expr);
+    Expression *expr = map.key_expr;
+
+    if (expr->type.IsTupleTy()) {
+      // We can only do this expression and size type manipulation
+      // if we have access to the underlying expressions in the tuple ast
+      // e.g. @map[(1, 2)] vs $a = (1, 2) @map[$a]
+      auto tuple_ast = dynamic_cast<Tuple *>(expr);
+      if (tuple_ast) {
+        is_tuple_ast = true;
+        std::vector<SizedType> elements;
+        Tuple *tuple = static_cast<Tuple *>(map.key_expr);
+        // NOLINTBEGIN(modernize-loop-convert)
+        for (unsigned int i = 0; i < tuple->elems.size(); i++) {
+          // NOLINTEND(modernize-loop-convert)
+          Visit(tuple->elems.at(i));
+          Expression *elem = tuple->elems.at(i);
+
+          // If elem type is none that means that the tuple contains some
+          // invalid cast (e.g., (0, (aaa)0)). In this case, skip this.
+          // Cast already emits the error.
+          if (elem->type.IsNoneTy() || elem->type.GetSize() == 0)
+            return;
+
+          // Insert a cast to 64 bits if needed by injecting
+          // a cast into the ast.
+          if (elem->type.IsIntTy() && elem->type.GetSize() < 8) {
+            Expression *cast = ctx_.make_node<Cast>(elem->type.IsSigned()
+                                                        ? CreateInt64()
+                                                        : CreateUInt64(),
+                                                    elem,
+                                                    map.loc);
+            Visit(cast);
+            tuple->elems[i] = cast;
+            elem = cast;
+          } else if (elem->type.IsPtrTy() && elem->type.IsCtxAccess()) {
+            // map functions only accepts a pointer to a element in the stack
+            LOG(ERROR, map.loc, err_) << "context cannot be used as a map key";
+          }
+
+          if (is_final_pass() && elem->type.IsNoneTy())
+            LOG(ERROR, elem->loc, err_)
+                << "Invalid expression for assignment: ";
 
-  for (unsigned int i = 0; i < map.vargs.size(); i++) {
-    Visit(map.vargs.at(i));
-    Expression *expr = map.vargs.at(i);
+          SizedType keytype = elem->type;
+          if (keytype.IsIntTy()) {
+            keytype = CreateUInt(keytype.GetSize() * 8);
```

Comment:
What does this do?

Reply:
This matches what we currently do for map keys, we increase the size of all ints to 8 bytes large - this probably needs further fixing though. https://github.com/bpftrace/bpftrace/blob/master/src/ast/passes/semantic_analyser.cpp#L1499

Reply:
But this just converts all int keys to uint, doesn't it? I was wondering why do we need that.

(the above link is probably outdated and points to some unrelated code so I cannot really compare)

Reply:
> But this just converts all int keys to uint, doesn't it? I was wondering why do we need that.

Yeah I was scratching my head about that too. I think our whole int conversion strategy needs re-evaluation but it also puts us in this situation of potentially breaking backward compatibility like with this recent change: https://github.com/bpftrace/bpftrace/pull/3416

I updated the link above.

Reply:
I see, ok, it's strange but it seems to be working so let's keep it for now.